src/frontend/src/optimizer/plan_node/derive.rs

```diff
@@ -120,5 +124,19 @@ pub(crate) fn derive_pk(
         in_order.insert(idx);
     }
 
+    // We need to ensure distribution key is part of pk.
+    // If it is not part of either stream_key or order_key,
+    // It must mean that it is only necessary for storage distribution.
+    // Such a case is rare, but it is possible,
+    // for example in the case of index created on singleton mv.
+    // In this case, we can simply append these columns to pk.
+    for &idx in input.distribution().dist_column_indices() {
+        if in_order.contains(idx) {
+            continue;
+        }
+        pk.push(ColumnOrder::new(idx, OrderType::ascending()));
+        in_order.insert(idx);
+    }
```

Comment:
BTW, this behavior could break the assumption (not sure whether still holds ATM) we made, i.e. the distribution key should be the prefix of the pk.  I remembered we have some discussions on this topic @st1page @fuyufjh

Reply:
My recollection is a bit hazy, but it seems to be for the performance consideration of batch scanning, the distribution key was restricted to the prefix of the primary key (pk), so that queries with a prefix can scan fewer partitions.

Reply:
ok, now I prefer to not append the distribution key to the end when distribution key is eliminated by functional dependency and we can use the original one instead.

Reply:
> ok, now I prefer to not append the distribution key to the end when distribution key is eliminated by functional dependency and we can use the original one instead. @kwannoel 

Hmm so just don't prune order key for indexes, since they are used as distribution key?

Reply:
Think the tricky thing is that distribution key is always a prefix of PK, and so is order key in certain cases.

There isn't a clear way to determine which should go before the other, unless we know the plans which required the distribution key and order key.

In the case of derive pk, we don't have this info. So seems hard to make a generalised decision. So perhaps better to just disable this case for index, seems to be the only one so far.

We can revisit if more edge cases show up.

Reply:
I am afraid of this PR eliminating the distribution key and then to ensure the distribution key is part of pk, we generate a new pk with the distribution key in the suffix. For this case, we can just give up the optimization and use the original derived pk.

Reply:
Fixed it by making distribution a constraint of minimizing order key. If pruning order key column also results in dist key column being pruned, we will not prune it.